src/mpid/ch4/netmod/ofi/ofi_types.h

```diff
@@ -507,6 +506,7 @@ typedef struct {
     struct fi_context context[MPIDI_OFI_CONTEXT_STRUCTS];       /* fixed field, do not move */
     int event_id;               /* fixed field, do not move */
     MPIR_Request *parent;       /* Parent request           */
+    void *pack_buffer;
```

Comment:
In this implementation, I tied the pack buffers to the "chunk" request object so they can be release as soon as they are completed. This requires enabling a completion for each pipeline chunk, which we only used before if using request-based RMA.

Another design might be to put the pack_buffers into a list on the window and release them when the window counter indicates all outstanding ops are complete. @yfguo I'm not sure if this relates to your AM pipeline framework, but I would appreciate your comments.

Reply:
@pavanbalaji your comments also welcome.

Reply:
@raffenet  I don't think you can wait till the entire operation is complete to release the buffers.  Consider the case the user wants to send a message that is larger than the entire pool size.  In that case, unless you release the buffers, some part of the transfer will never get initiated.

Reply:
It shouldn't need to wait for the entire operation to complete to release the buffers, only the in-flight chunks. The deferred queue tracks whether an operation still has remaining work, so the op can be restarted once more resources are available.

Reply:
The benefit is that we do not need to add an ofi completion for the pipeline chunk.

Reply:
Is the concern with having individual completions that of the performance cost or of the code complexity?  The performance cost should be minimal because this is only for large messages and also only for ones that need temporary buffering.  I don't fully understand the code complexity concerns.

Reply:
I was somewhat concerned about the completion cost, but you are right we are doing data packing here so that cost is likely insignificant.

My other concern is availability of pack buffers. In this version, the pack buffers become available as soon as the chunk is completed, no matter how progress is made. If I attach all the pack buffers to a window and only free them when making progress on *that* window, then an application could be somewhat limited in terms of buffer availability. So I think I've talked myself into sticking with this version. I'm going to split and clean up the last commit a little, but this is mostly done.

Reply:
I think also if pipelining MPI_GET operations (only MPI_PUT in this PR so far), I'm going to need the chunk context object to store the offset value for unpacking.